compiler/resolution/resolveFunction.cpp

```diff
@@ -574,6 +588,233 @@ static void markTypesWithDefaultInitEqOrAssign(FnSymbol* fn) {
+
+    VarSymbol* toTmp = newTemp("_check_to", toType);
+    VarSymbol* fromTmp = newTemp("_check_from", fromType);
+    CallExpr* c = new CallExpr(astrSassign, toTmp, fromTmp);
+    block->insertAtTail(new DefExpr(toTmp));
+    block->insertAtTail(new DefExpr(fromTmp));
+    block->insertAtTail(c);
+    tryResolveCall(c);
+    block->remove();
+    // update our local variable
+    // (above resolve call should have updated the map)
+    have.assign = c->resolvedFunction();
+  }
+
+  if (checkInitEq && have.initEq == NULL) {
+    // also resolve init=
+    BlockStmt* block = new BlockStmt(BLOCK_SCOPELESS);
+    forCall->insertBefore(block);
+
+    VarSymbol* toTmp = newTemp("_check_to", toType);
+    VarSymbol* fromTmp = newTemp("_check_from", fromType);
+    CallExpr* c = new CallExpr(astrInitEquals, gMethodToken, toTmp, fromTmp);
+    block->insertAtTail(new DefExpr(toTmp));
+    block->insertAtTail(new DefExpr(fromTmp));
+    block->insertAtTail(c);
+    tryResolveCall(c);
+    block->remove();
+    // update our local variable
+    // (above resolve call should have updated the map)
+    have.initEq = c->resolvedFunction();
+  }
+
+  if (checkCast && have.cast == NULL) {
+    // also resolve cast
+    BlockStmt* block = new BlockStmt(BLOCK_SCOPELESS);
+    forCall->insertBefore(block);
+
+    VarSymbol* fromTmp = newTemp("_check_from", fromType);
+    CallExpr* c = new CallExpr(astr_cast, toType->symbol, fromTmp);
+    block->insertAtTail(new DefExpr(fromTmp));
+    block->insertAtTail(c);
+    tryResolveCall(c);
+    block->remove();
+    // update our local variable
+    // (above resolve call should have updated the map)
+    have.cast = c->resolvedFunction();
+  }
+
+  // Now, error if an implied function was not found.
+  bool error = false;
+  if (checkAssign && have.assign == NULL) {
+    USR_FATAL_CONT(forCall, "an = overload setting '%s' from '%s' is missing",
+                            toString(toType), toString(fromType));
+    error = true;
+  }
+  if (checkInitEq && have.initEq == NULL) {
+    USR_FATAL_CONT(forCall, "an init= initializing '%s' from '%s' is missing",
+                            toString(toType), toString(fromType));
+    error = true;
+  }
+  if (checkCast && have.cast == NULL) {
+    USR_FATAL_CONT(forCall, "a cast creating '%s' from '%s' is missing",
+                            toString(toType), toString(fromType));
+    error = true;
+  }
+
+  if (error) {
+    if (implicitConverts) {
+      USR_PRINT("implicit conversion is available between these two types");
+    }
+    if (have.assign != NULL) {
+      USR_PRINT(have.assign->defPoint,
+                "assignment is defined here between these two types");
+    }
+    if (have.initEq != NULL) {
+      USR_PRINT(have.initEq->defPoint,
+                "init= is defined here between these two types");
+    }
+    if (have.cast != NULL) {
+      USR_PRINT(have.cast->defPoint,
+                "cast is defined here between these two types");
```

Comment:
I think it would be more clear what these messages have to do with the error if they all started with "expected because "